torch/optim/swa_utils.py

```diff
@@ -121,23 +150,26 @@ def update_parameters(self, model):
             itertools.chain(model.parameters(), model.buffers())
             if self.use_buffers else model.parameters()
         )
+        self_param_detached = []
+        model_param_detached = []
         for p_swa, p_model in zip(self_param, model_param):
             device = p_swa.device
             p_model_ = p_model.detach().to(device)
+            self_param_detached.append(p_swa.detach())
+            model_param_detached.append(p_model_)
             if self.n_averaged == 0:
                 p_swa.detach().copy_(p_model_)
+
+        if self.n_averaged > 0:
+            if self.multi_avg_fn is not None or self.avg_fn is None:
```

Comment:
This would mean the default (if no avg_fn nor multi_avg_fn were provided) would call the path that would call torch.foreach, which is only defined for CUDA/CPU (and so tensors like mps tensors or any other custom device tensors would fail).

Could we have it so that the default only calls the multi_avg_fn if they're cuda tensors? This will require some more branching here.

Reply:
looks good